packages/react/src/ToggleSwitch/ToggleSwitch.tsx

```diff
@@ -223,27 +243,64 @@ const ToggleSwitch = React.forwardRef<HTMLButtonElement, React.PropsWithChildren
       onClick,
       size = 'medium',
       statusLabelPosition = 'start',
+      loadingLabelDelay = 2000,
+      loadingLabel = 'Loading',
       sx: sxProp,
       ...rest
     } = props
     const isControlled = typeof checked !== 'undefined'
     const [isOn, setIsOn] = useProvidedStateOrCreate<boolean>(checked, onChange, Boolean(defaultChecked))
     const acceptsInteraction = !disabled && !loading
+
+    const [loadingLabelTimeoutId, setLoadingLabelTimeoutId] = React.useState<number | null>(null)
+    const [isLoadingLabelVisible, setIsLoadingLabelVisible] = React.useState(false)
+    const loadingLabelId = useId('loadingLabel')
+
+    const resetLoadingLabel = useCallback(() => {
+      if (loadingLabelTimeoutId) {
+        clearTimeout(loadingLabelTimeoutId)
+        setLoadingLabelTimeoutId(null)
+      }
+
+      if (isLoadingLabelVisible) {
+        setIsLoadingLabelVisible(false)
+      }
+    }, [loadingLabelTimeoutId, isLoadingLabelVisible])
+
     const handleToggleClick: MouseEventHandler = useCallback(
       e => {
+        if (disabled || loading) return
+
         if (!isControlled) {
           setIsOn(!isOn)
+          resetLoadingLabel()
         }
         onClick && onClick(e)
       },
-      [onClick, isControlled, isOn, setIsOn],
+      [disabled, isControlled, loading, onClick, setIsOn, isOn, resetLoadingLabel],
     )
 
     useEffect(() => {
-      if (onChange && isControlled) {
+      if (onChange && isControlled && !disabled) {
         onChange(Boolean(checked))
       }
-    }, [onChange, checked, isControlled])
+    }, [onChange, checked, isControlled, disabled])
+
+    if (loading) {
+      if (!loadingLabelTimeoutId) {
+        setLoadingLabelTimeoutId(
+          setTimeout(() => {
+            setLoadingLabelTimeoutId(null)
+            setIsLoadingLabelVisible(true)
+          }, loadingLabelDelay) as unknown as number,
+        )
+      }
```

Comment:
I think this could be considered a side-effect during render since we're creating a side-effect (the timeout) outside of an effect. I think we'll need this code to be wrapped in an effect (or triggered by an event handler) to prevent this

Reply:
Agreed! Most of the functionality should now be in an effect!